### ISSUE TITLE
sync from blob to local actually copies everything again and again?

### ISSUE THREAD

**Issue by ppanyukov** [Azure/azure-storage-azcopy#220]
### Which version of the AzCopy was used? 
azcopy_linux_amd64_10.0.7

### Which platform are you using? (ex: Windows, Mac, Linux)
Linux

### What command did you run?

```
/azcopy_linux_amd64_10.0.7/azcopy sync \
   "https://acc.blob.core.windows.net/container?SAS"  \
   ./localdir \
   --recursive=true \
   --log-level INFO
```

### What problem was encountered?

All is good upon first sync.

No changes in blob storage or local directory.

First, running subsequent sync give this message:

```
Sync has enumerated 13 files to delete locally. 
Do you want to delete these files ? 
Please confirm with y/n:
```

Very confusing! Why does it want to delete local files?

Second, after answering "N" to the above question, even though blob storage and local directory are fully in sync, the sync command still seems to pull down *everything* from blob storage.  This is confirmed by looking at the logs: it makes requests to fetch every single blob again.

The last modified time stamps seem to agree between blob storage and local file system.

### Have you found a mitigation/solution?

If I can make a suggestion, the above behaviour is odd.  I'd rather expected it to be similar to `rsync` -- only grab files which are different. For example, a simple variant of this: 

```
if modified date or size are different; then
   fetch
else
   skip
fi
```

A more extended version might be to check MD5 checksums if they are set on the blob objects in addition/instead of the above.

If this is the expected behavior, then this doesn't work for me.

**Comment by zezha-msft** [comment]
Hi @ppanyukov, thanks for reaching out! And sorry for the inconvenience. 

Sync tries to replicate the source's content to the destination. This means that if there are extra files at the destination that are not present at the source, we'd prompt the user to see if they should be removed. This behavior is great if you had a folder containing webpages and stylesheets, and you sync this folder to the $web container to host a static website in Azure Storage; if you deleted a html file from your local folder, it'd make sense to also delete it at the $web container the next time you do a sync, right?

Btw sync was rewritten and this bug of retransferring the same files should be fixed. It will go out with the next release.

**Comment by zezha-msft** [comment]
Here is a [brief writeup](https://github.com/Azure/azure-storage-azcopy/wiki/Sync-command-guide) on how the sync command compares source and destination. We'd really appreciate your feedback on it. Thank you!!

**Comment by williexu** [comment]
@zezha-msft I just ran the sync command while the source and destination were at parity and was presented with the same message:
`Sync has enumerated 6 files to delete from destination. Do you want to delete these files ? Please confirm with y/n:`
After confirming with Y, I found that ALL of my files at the destination were deleted, despite being all present in the source. Seems like a bug.
The following output from a CLI command uses the flags `--recursive` and `--force`.
```
(envnew) C:\Users\wilx\azure-cli>az storage azcopy blob sync --destination container2 --source ./garbage --account-name wilxstorage --auth-mode login
AZCOPY_OAUTH_TOKEN_INFO is set.
Using OAuth token for authentication.

Job b3077215-fb4b-cb4a-53b8-36bbe4c6d8a9 has started

Log file is located at: C:\Users\wilx/.azcopy/b3077215-fb4b-cb4a-53b8-36bbe4c6d8a9.log
0 File Scanned at Source, 0 Files Scanned at Destination


Job b3077215-fb4b-cb4a-53b8-36bbe4c6d8a9 summary
Elapsed Time (Minutes): 0.0334
Total Number Of Copy Transfers: 6
Total Number Of Delete Transfers: 0
Number of Copy Transfers Completed: 6
Number of Copy Transfers Failed: 0
Number of Delete Transfers Completed: 0
Number of Delete Transfers Failed: 0
Final Job Status: Completed


(envnew) C:\Users\wilx\azure-cli>az storage azcopy blob sync --destination container2 --source ./garbage --account-name wilxstorage --auth-mode login
AZCOPY_OAUTH_TOKEN_INFO is set.
Using OAuth token for authentication.

Job ebc75027-dc8d-3b44-7d64-8271a6187e31 has started

Log file is located at: C:\Users\wilx/.azcopy/ebc75027-dc8d-3b44-7d64-8271a6187e31.log
6 Done, 0 Failed, 6 Pending, 12 Total, 2-sec Throughput (Mb/s): 0.0514


Job ebc75027-dc8d-3b44-7d64-8271a6187e31 summary
Elapsed Time (Minutes): 0.0667
Total Number Of Copy Transfers: 6
Total Number Of Delete Transfers: 6
Number of Copy Transfers Completed: 6
Number of Copy Transfers Failed: 0
Number of Delete Transfers Completed: 6
Number of Delete Transfers Failed: 0
Final Job Status: Completed
```
Somehow on the second iteration (which should do nothing), all the files are deleted.

**Comment by zezha-msft** [comment]
@williexu yes, the previous implementation had a lot of bugs. Rest assured that I've rewritten the sync command and implemented extensive tests to guarantee the quality.

**Comment by seguler** [comment]
I confirm - tested the private drop and all works fine in the next release :)

**Comment by Kapanther** [comment]
@zezha-msft just read your write up 
I can see why item 2 would be ideal behaviour
> the same file also exists at the destination, but its last modified time is earlier than the source file
but obviously a flag to overwrite regardless of date modified something like --mirrordest would probably also be very useful. But i think you have this covered in 10.0.8.

**Comment by zezha-msft** [comment]
Hi @Kapanther, are you talking about the [wiki](https://github.com/Azure/azure-storage-azcopy/wiki/Sync-command-guide)? 😄 

If the user wanted to overwrite regardless, then they should use `copy` with `overwrite` flag set to true. Please let us know if you think we are missing a scenario.

**Comment by Kapanther** [comment]
@zezha-msft possibly you are...

If I use `copy` it will not remove files in the destination that have been created. Thus the necessity of sync

It should be possibly to perform a "one-way sync", where the destination mirrors the source. And will wipe any newer or non-existent files in the destination container as well. (this is currently not possible in azcopy)

Similar to the behaviour of of using `robocopy <source> <dest> /MIR /PURGE`. Where;
`/MIR` mirrors the directory tree and files from the source to the dest
`/PURGE` removes files from the destination that don't exist on the source.

maybe a flag `--purge` flag makes more sense?

**Comment by zezha-msft** [comment]
Hi @Kapanther, if I'm understanding you correctly, you want a `--check-lmt` flag or something similar to indicate whether sync should compare last modified times, right? If the flag value is false, we'd overwrite all destination files without looking at their lmts. If the flag value is true, it's the current sync behavior.

We do have a new flag called `delete-destination` that resembles `--purge`. If true, it would purge; if false; it would leave the extra files at destination alone; if prompt, we'd ask the user to see what to do with extra files once we discover the first one. The default value is false.

**Comment by Kapanther** [comment]
@zezha-msft exactly right... `--check-lmt` is exactly what is needed.  Otherwise newer files cant be handled by azcopy.

**Comment by zezha-msft** [comment]
Hi @artemuwka please take a look at @Kapanther's suggestion. Thanks!

**Comment by zezha-msft** [comment]
Hi @ppanyukov, sorry for the inconvenience, and I really appreciate your patience!

I've rewritten the sync command and released it in 10.0.8. This issue should be fixed. Please let us know if you still encounter any problem.

Here is a [wiki](https://github.com/Azure/azure-storage-azcopy/wiki/Sync-command-guide) explaining how the new sync command works. Please feel free to reach out if you have any question or concern.

**Comment by zezha-msft** [comment]
Hi @Kapanther, sorry I need to close this issue to mark it as done. Please feel free to open another issue for you suggestion.
